### PR TITLE
Updated the specification file and reformated all the patches incrementaly

### PR DESCRIPTION
**Pull request by xavierlacot**
Hi, I updated the openapi spec file and reapplied/updated/cleanup all the patches. Please note that I did not check in Slack doc if the patches are useful or not.

I noticed that some of them have been merged in the official Slack openapi file, which lead me to remove the patch 9 and slightly reduce the size of the patch 10.

### CODE REVIEW

**Comment by damienalexandre** [comment]
Looks good to me, the only change I found strange is:

```diff
-        $optionsResolver->setRequired([]);
+        $optionsResolver->setRequired(['token']);
```

On almost all endpoints it seems (not all, like ConversationsCreate.php).

It's true that the token is required: https://api.slack.com/methods/emoji.list#arg_token but that's supposed to be handled by our Client Factory: https://github.com/jolicode/slack-php-api/blob/28967f79835df4266da5d18979b0a78986a7d62f/src/ClientFactory.php#L42

You should never have to send to token yourself and now it's mandatory in the endpoint options.

That's not your fault, the spec changed:

## Before

```
                "operationId": "emoji_list",
                "parameters": [
                    {
                        "description": "Authentication token. Requires scope: `emoji:read`",
                        "in": "query",
                        "name": "token",
                        "type": "string"
                    }
                ],
```

## Now

```
                "operationId": "admin_emoji_list",
                "parameters": [
                    {
                        "description": "Set `cursor` to `next_cursor` returned by the previous call to list items in the next page",
                        "in": "query",
                        "name": "cursor",
                        "type": "string"
                    },
                    {
                        "description": "Authentication token. Requires scope: `admin.teams:read`",
                        "in": "query",
                        "name": "token",
                        "required": true,
                        "type": "string"
                    },
                    {
                        "description": "The maximum number of items to return. Must be between 1 - 1000 both inclusive.",
                        "in": "query",
                        "name": "limit",
                        "type": "integer"
                    }
                ],
```

## What the documentation says

From https://api.slack.com/web 

> Authenticate your Web API requests by providing a bearer token, which identifies a single user, bot user, or workspace-application relationship.
> Register your application with Slack to obtain credentials for use with our OAuth 2.0 implementation, which allows you to negotiate tokens on behalf of users and workspaces.
> **We prefer tokens to be sent in the Authorization HTTP header of your outbound requests. However, you may also pass tokens in all Web API calls as a parameter called token.**

## How to fix?

We have two choices:

- :a: accept the PR as is, breaking our promise on the Client Factory, adding a "token" parameters "sometimes"
- :b: add a new patch that remove the "required" part of the "token" parameter to allow using the Client as easily as before

I'm in favor of :b: - what do you think?

**Comment by xavierlacot** [comment]
Yes, :b: seems a better solution (from a DX point of vue). I am going to work on the changes.

**Comment by xavierlacot** [comment]
Another point, which is quite related: it was a pain to reformat the patches. Very often, the `patch` tool looses himself trying to figure out which line to change... and it fails, applying changes where is shouldn't, etc. This makes the current process hardly maintainable in the long term, I would say.

However, there are two things we could do to make patches more "robust":

1. sort the openapi spec file by keys, recursively. This will not change the meaning of the spec, and will ensure that the keys always come in the same order, even if Slack decides to wrangle the file paths order.
2. add more context in patches. At the moment, lines like [10-more-completeness.patch#L96-L97](https://github.com/jolicode/slack-php-api/blob/master/resources/patches/10-more-completeness.patch#L96-L97) are not applied at the right position, because the patch tool misses context. We can add more context lines above and below a diff by passing an additional `-U15` parameter to `git diff`, which should help. Combined with the first point, it should make the patching process more robust, I believe.

If you agree, I would add in this pull request:
 * a CLI command that would download the original vanilla spec, turn into an alphabtically sorted equivalent, then apply the patches and generate the `slack-openapi-patched.json` file
 * a new "reset" patch equivalent to the current patches, but applicable on the sorted vanilla spec.
 * a cli command that would regenerate this patch file base on the current diff between `slack-openapi-patched.json` and `slack-openapi.json`. This will give us a single patch file, but versionned, which will allow to track the patch changes when Slack updates the vanilla spec.

This way, the next time a patch has to be generated:
1. the developer updates the vanilla spec using a CLI command
2. the developer makes his changes in the generated patched spec file (which is sorted by keys)
3. the developer regenerates the patch using a CLI command
4. the developer generates the SDK and commits all the changes

What would you thing about this?

**Comment by damienalexandre** [comment]
I like this :+1: but I fear having only one patch file will make it hard to read "what is changed from the official spec". Having multiple files allowed us to be incremental - but as you said it's a pain to update. 
:+1: for your proposal.

We can also switch to the new "no example" specification: https://github.com/slackapi/slack-api-specs/blob/master/web-api/slack_web_openapi_v2_without_examples.json - that way exemples will not get in our way when changed / edited.

**Comment by xavierlacot** [comment]
Yay :muscle: 
@damienalexandre and @pyrech can you take a look at it?

The commit f6d15613983f0c49f5afa3fcffe0482d436d3cc0 exactly shows what we earn when a change is made in te spec: mainly a diff in the patches file, which is quite readable. Of course, we'll still have a detailed review to undergo if Slack decides to change all the format if the spec, but it should a lot more stable now.

File: doc/updating-sdk.md [hunk 1]
```diff
@@ -1,53 +1,48 @@
...
-Run the following command to retrieve the latest API spec:
+Sometimes, the API specification is not complete or incorrect. Slack is aware of these problems,
+but we just can't wait months for an update. Most of the time, examples inside the spec are
+correct, and we just miss properties in the response's schema definition.
 
-```bash
-curl https://api.slack.com/specs/openapi/v2/slack_web.json -o resources/slack-openapi.json
-```
+This is the reason why we maintain a patch for the official schema, in order to reflect the
+contents of the official documentation.
```
Comment (damienalexandre) [inline doc/updating-sdk.md:10]:
```suggestion
This is the reason why we maintain a patched version of the official schema, following both the documentation and PHP specific requirements.
```

File: doc/updating-sdk.md [hunk 2]
```diff
@@ -1,53 +1,48 @@
...
 
-:warning: This may invalidate our versioned patches!
+in order to download the latest API specification and apply the patches, you may run this command:
```
Comment (damienalexandre) [inline doc/updating-sdk.md:12]:
```suggestion
In order to download the latest API specification and apply the patches, you may run this command:
```

File: doc/updating-sdk.md [hunk 3]
```diff
@@ -1,53 +1,48 @@
...
 
-Most of the times, examples inside the spec are correct and we just miss
-properties in the response's schema definition.
+This command sequentially runs the following operations:
 
-So we patch the schema with what we can see in the official documentation via
-the patch command.
+1. it downloads and saves the last [official Slack OpenAPI specification file](https://api.slack.com/specs/openapi/v2/slack_web.json) in [resources/slack-openapi.json](https://github.com/jolicode/slack-php-api/blob/master/resources/slack-openapi.json)
+2. it recursively sorts the json keys and saves this intermediate version in [resources/slack-openapi-sorted.json](https://github.com/jolicode/slack-php-api/blob/master/resources/slack-openapi-sorted.json)
+3. it applies [our patches](https://github.com/jolicode/slack-php-api/blob/master/resources/slack-openapi-sorted.patch) and stores the resulting specification file in [resources/slack-openapi-patched.json](https://github.com/jolicode/slack-php-api/blob/master/resources/slack-openapi-patched.json)
```
Comment (damienalexandre) [inline doc/updating-sdk.md:22]:
```suggestion
1. Downloads and saves the current [official Slack OpenAPI specification file](https://api.slack.com/specs/openapi/v2/slack_web.json) in [resources/slack-openapi.json](https://github.com/jolicode/slack-php-api/blob/master/resources/slack-openapi.json)
2. Recursively sorts the JSON keys and saves this intermediate version in [resources/slack-openapi-sorted.json](https://github.com/jolicode/slack-php-api/blob/master/resources/slack-openapi-sorted.json)
3. Applies [our patches](https://github.com/jolicode/slack-php-api/blob/master/resources/slack-openapi-sorted.patch) and stores the resulting specification file in [resources/slack-openapi-patched.json](https://github.com/jolicode/slack-php-api/blob/master/resources/slack-openapi-patched.json)
```

File: doc/updating-sdk.md [hunk 4]
```diff
@@ -1,53 +1,48 @@
...
 
-## Regenerate the SDk
+The patches in this repository may not fix all the specification issues. Should you discover
+a new specification bug, here is the process to propose changes:
 
-When the versioned spec has been updated, we need to run Jane to regenerate the
-SDK:
+1. edit the `resources/slack-openapi-patched.json` file as you wish
+2. test the updated SDK (see above)
+3. when satisfied with the result, run this command:
```
Comment (damienalexandre) [inline doc/updating-sdk.md:42]:
```suggestion
1. Edit the `resources/slack-openapi-patched.json` file as you wish
2. Update the SDK (see above) and test the resulting code
3. When satisfied with the result, run this command:
```

File: generated/Endpoint/ChatUpdate.php [hunk 5]
```diff
@@ -66,16 +66,16 @@ public function getExtraHeaders(): array
...
         $optionsResolver = parent::getFormOptionsResolver();
-        $optionsResolver->setDefined(['blocks', 'attachments', 'text', 'ts', 'parse', 'as_user', 'link_names', 'channel']);
-        $optionsResolver->setRequired([]);
+        $optionsResolver->setDefined(['blocks', 'attachments', 'as_user', 'ts', 'parse', 'text', 'link_names', 'channel']);
+        $optionsResolver->setRequired(['ts', 'channel']);
         $optionsResolver->setDefaults([]);
         $optionsResolver->setAllowedTypes('blocks', ['string']);
         $optionsResolver->setAllowedTypes('attachments', ['string']);
-        $optionsResolver->setAllowedTypes('text', ['string']);
+        $optionsResolver->setAllowedTypes('as_user', ['string']);
```
Comment (damienalexandre) [inline generated/Endpoint/ChatUpdate.php:74]: I don't understand why Slack changed this boolean to a string... But let's see :grimacing:

**Comment by xavierlacot** [comment]
:pray: thanks for the feedback on the documentation

**Comment by damienalexandre** [comment]
Tests are green, let's move this to master!

Release is postponed as we are waiting for another improvement (Jane upgrade) :+1:

**Pull request merged**